zerver/lib/message_cache.py

```diff
@@ -219,6 +222,16 @@ def finalize_payload(
         if not skip_copy:
             obj = copy.copy(obj)
 
+        # Compatibility code to change topic="" to topic="general chat"
+        # for clients with no UI support for empty topic name.
+        # TODO: Improve this comment.
+        if (
+            obj["recipient_type"] == Recipient.STREAM
+            and obj["subject"] == ""
+            and not allow_empty_topic_name
+        ):
+            obj["subject"] = "general chat"
```

Comment:
So I had a thought on this -- we could perhaps have `topic_name` be available and have a value of `""`, and a legacy/deprecated `"subject"` field present whose value is `"general chat"` here. 

The idea being that a modern client would ignore `subject` and just parse the `topic_name` field, and thus we'd be able to remove the legacy `subject` name along with this transition. And the `allow_empty_topic_name` client capability could then just have a meaning of "skip sending the legacy `subject` field". 

Worth discussing in the API design channel, but this might be a pretty compelling way to move forward.

Reply:
Having now read the rest of the branch, I think the essential idea of that proposal is a good idea, but probably something we can do in a follow-up. Basically, we'd introduce a `topic_name` field on `Message` that equals `subject`, except that it doesn't support `!allow_empty_topic_name`, under the assumptions that clients that migrate to reading it should post-date that API change anyway. And in doing so, we'd at least set ourselves up for the final end of the original "subject" name for topics (we renamed it in the UI in like 2013).